chunky/src/java/se/llbit/math/BigPackedOctree.java

```diff
@@ -0,0 +1,773 @@
+    if (invDirX >= 0) {
+      tMin = -ray.o.x * invDirX;
+      tMax = (octree_size - ray.o.x) * invDirX;
+
+      nx = -1;
+      ny = nz = 0;
+    } else {
+      tMin = (octree_size - ray.o.x) * invDirX;
+      tMax = -ray.o.x * invDirX;
+
+      nx = 1;
+      ny = nz = 0;
+    }
+
+    double tYMin, tYMax;
+    double invDirY = 1 / ray.d.y;
+    if (invDirY >= 0) {
+      tYMin = -ray.o.y * invDirY;
+      tYMax = (octree_size - ray.o.y) * invDirY;
+    } else {
+      tYMin = (octree_size - ray.o.y) * invDirY;
+      tYMax = -ray.o.y * invDirY;
+    }
+
+    if ((tMin > tYMax) || (tYMin > tMax))
+      return false;
+
+    if (tYMin > tMin) {
+      tMin = tYMin;
+
+      ny = -FastMath.signum(ray.d.y);
+      nx = nz = 0;
+    }
+
+    if (tYMax < tMax)
+      tMax = tYMax;
+
+    double tZMin, tZMax;
+    double invDirZ = 1 / ray.d.z;
+    if (invDirZ >= 0) {
+      tZMin = -ray.o.z * invDirZ;
+      tZMax = (octree_size - ray.o.z) * invDirZ;
+    } else {
+      tZMin = (octree_size - ray.o.z) * invDirZ;
+      tZMax = -ray.o.z * invDirZ;
+    }
+
+    if ((tMin > tZMax) || (tZMin > tMax))
+      return false;
+
+    if (tZMin > tMin) {
+      tMin = tZMin;
+
+      nz = -FastMath.signum(ray.d.z);
+      nx = ny = 0;
+    }
+
+    if (tMin < 0)
+      return false;
+
+    ray.o.scaleAdd(tMin, ray.d);
+    ray.n.set(nx, ny, nz);
+    ray.distance += tMin;
+    return true;
+  }
+
+  /**
+   *  {@inheritDoc}
+   */
+  @Override
+  public boolean enterBlock(Scene scene, Ray ray, BlockPalette palette) {
+    if (!isInside(ray.o) && !enterOctree(ray))
+      return false;
+
+    // Marching is done in a top-down fashion: at each step, the octree is descended from the root to find the leaf
+    // node the ray is in. Terminating the march is then decided based on the block type in that leaf node. Finally the
+    // ray is advanced to the boundary of the current leaf node and the next, ready for the next iteration.
+    while (true) {
+      // Add small offset past the intersection to avoid
+      // recursion to the same octree node!
+      int x = (int) QuickMath.floor(ray.o.x + ray.d.x * Ray.OFFSET);
+      int y = (int) QuickMath.floor(ray.o.y + ray.d.y * Ray.OFFSET);
+      int z = (int) QuickMath.floor(ray.o.z + ray.d.z * Ray.OFFSET);
+
+      int lx = x >>> depth;
+      int ly = y >>> depth;
+      int lz = z >>> depth;
+
+      if (lx != 0 || ly != 0 || lz != 0)
+          return false; // outside of octree!
+
+      // Descend the tree to find the current leaf node
+      int level = depth;
+      long nodeIndex = 0;
+      while(getAt(nodeIndex) > 0) {
+//        System.out.printf("%d at level %d\n", getAt(nodeIndex), level);
+        level -= 1;
+        lx = x >>> level;
+        ly = y >>> level;
+        lz = z >>> level;
+        nodeIndex = getAt(nodeIndex) + (((lx & 1) << 2) | ((ly & 1) << 1) | (lz & 1));
+      }
+//      System.out.printf("%d at level %d\n", getAt(nodeIndex), level);
+
+      //System.exit(0);
```

Comment:
The comments and old debug output can probably be removed.

Reply:
oops ^

Reply:
Fixed